common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java

```diff
@@ -1232,22 +1255,62 @@ public static UTF8String trimLeft(
     // Return the substring from the calculated position until the end of the string.
     return UTF8String.fromString(src.substring(charIndex));
   }
+  /**
+   * Trims the `srcString` string from the right side using the specified `trimString` characters,
+   * with respect to the UTF8_BINARY trim collation. For UTF8_BINARY trim collation, the method has
+   * one special case to cover with respect to trimRight function for regular UTF8_Binary collation.
+   * Trailing spaces should be ignored in case of trim collation (rtrim for example) and if
+   * trimString does not contain spaces. In this case, the method trims the string from the right
+   * and after call of regular trim functions returns back trimmed spaces as those should not get
+   * removed.
+   * @param srcString the input string to be trimmed from the right end of the string
+   * @param trimString the trim string characters to trim
+   * @param collationId the collation ID to use for string trim
+   * @return the trimmed string (for UTF8_LCASE collation)
+   */
+  public static UTF8String binaryTrimRight(
+      final UTF8String srcString,
+      final UTF8String trimString,
+      final int collationId) {
+    boolean ignoreSpaces = !trimString.contains(UTF8String.SPACE_UTF8) &&
+      CollationFactory.ignoresSpacesInTrimFunctions(collationId, false, true);
+
+    if (!ignoreSpaces) {
+      return srcString.trimRight(trimString);
+    }
+    int lastNonSpacePosition = srcString.numChars();
+    while (lastNonSpacePosition > 0 &&
+            srcString.getByte(lastNonSpacePosition - 1) == 0x20) {
+      --lastNonSpacePosition;
+    }
+
+    StringBuilder sb = new StringBuilder();
+    sb.append(srcString.trimRight().trimRight(trimString));
+    if (lastNonSpacePosition != srcString.numChars()) {
+      sb.append(srcString.substring(lastNonSpacePosition, srcString.numChars()));
+    }
```

Comment:
so here we have
- 2 UTF8String.trim calls, which will all allocate new UTF8Strings by calling `copyUTF8String`
- StringBuilder allocation
- additional UTF8String.substring call, which will allocate yet another UTF8String
- returning the final result, which first allocates a java String, and then another UTF8String

Reply:
I think we should try to do better...